File: android/app/src/main/kotlin/org/gaindrive/android/data/MusicRoots.kt
package org.gaindrive.android.data

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.gaindrive.android.data.model.MusicRoot
import org.gaindrive.android.data.model.ServerId
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Each server's configured roots, fetched once and kept for the session.
 *
 * Roots are configuration: they change when someone edits the server, not while
 * someone is browsing it. Re-asking on every chip press would put a request in
 * front of a control that should feel instant.
 *
 * It lives here rather than inside `LibraryRepository` — where it began — for
 * one structural reason: the repository injects `ServerRegistry`, so the
 * registry cannot reach back into it to invalidate a server whose browse mode
 * has just changed. A singleton of its own is injectable by both, and it is the
 * same shape as the other two per-server caches the registry already clears,
 * `SubsonicClientFactory.forget` and `LocalLibrary.forgetServer`.
 */
@Singleton
class MusicRoots @Inject constructor() {

	private val cache = mutableMapOf<ServerId, List<MusicRoot>>()
	private val mutex = Mutex()

	/**
	 * The roots of [server], fetching them with [load] the first time.
	 *
	 * A failure caches nothing, so a server that was briefly unreachable is
	 * asked again rather than being remembered as having no roots at all — which
	 * would quietly hold its chips back for the rest of the session.
	 */
	suspend fun of(server: ServerId, load: suspend () -> List<MusicRoot>): List<MusicRoot> {
		cache[server]?.let { return it }
		return mutex.withLock {
			cache[server] ?: load().also { cache[server] = it }
		}
	}

	fun forget(server: ServerId) {
		cache.remove(server)
	}
}

File: ios/GainDrive/Data/MusicRoots.swift
//	GainDrive for iOS
//	Copyright (C) 2026 Kasper Peeters
//
//	This file is part of GainDrive, distributed under the GNU General
//	Public License version 3 or later, with the Application Distribution
//	Exception in ios/LICENSE. See LICENSE at the repository root for the
//	full text of the GPL.

import Foundation

/// Each server's configured roots, fetched once and kept for the session.
///
/// Roots are **configuration**: they change when someone edits the server, not
/// while someone is browsing it. Re-asking on every chip press would put a
/// request in front of a control that should feel instant.
///
/// A store rather than a fetcher, unlike `Accounts`: the caller does the
/// request and only hands back what succeeded. That is what keeps the rule
/// simple - **a failure caches nothing**, so a server that was briefly
/// unreachable is asked again rather than remembered as having no roots at all,
/// which would quietly hold its chips back for the rest of the session.
///
/// It carries no in-flight deduplication, and `Accounts` does. The difference
/// is what waits on them: the first stream URL of a session blocks on an
/// account's ceiling, where two concurrent asks would be two requests before
/// the first note. Roots are read by a fan-out that already asks each server
/// once, so the worst case is one extra `getMusicFolders` - cheap, idempotent,
/// and not worth the machinery.
actor MusicRoots {
	private var cache: [ServerId: [MusicRoot]] = [:]

	func cached(_ server: ServerId) -> [MusicRoot]? {
		cache[server]
	}

	func store(_ roots: [MusicRoot], for server: ServerId) {
		cache[server] = roots
	}

	func forget(_ server: ServerId) {
		cache[server] = nil
	}
}
